File: src/main/java/com/nc5/generator/generator/ClientGenerator.java
package com.nc5.generator.generator;

import com.nc5.generator.config.BillConfig;
import com.nc5.generator.template.TemplateContext;
import com.nc5.generator.template.TemplateEngine;
import com.nc5.generator.template.TemplateSelector;
import org.apache.commons.io.FileUtils;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

import java.io.File;
import java.io.IOException;

/**
 * 客户端代码生成器
 */
public class ClientGenerator {

    private static final Logger logger = LoggerFactory.getLogger(ClientGenerator.class);

    private TemplateEngine templateEngine;

    private String outputDir;

    public ClientGenerator(String outputDir) {
        this.templateEngine = new TemplateEngine();
        this.outputDir = outputDir;
    }

    /**
     * 生成所有客户端代码
     */
    public void generate(BillConfig billConfig) throws IOException {
        logger.info("开始生成客户端代码");

        // 生成Controller
        generateController(billConfig);

        // 生成IPrivateBtn
        generateIPrivateBtn(billConfig);

        // 生成ClientUI
        generateClientUI(billConfig);

        // 生成BusinessAction
        generateBusinessAction(billConfig);

        // 生成Delegator
        generateDelegator(billConfig);

        // 生成EventHandler
        generateEventHandler(billConfig);

        // 生成RefModel
        generateRefModel(billConfig);

        logger.info("客户端代码生成完成");
    }

    /**
     * 生成Controller
     */
    private void generateController(BillConfig billConfig) throws IOException {
        TemplateContext context = new TemplateContext();
        context.setBillConfig(billConfig);

        // 使用模板选择器获取模板路径
        String templatePath = TemplateSelector.getControllerTemplate(billConfig);
        String content = templateEngine.render(templatePath, context);

        // 输出路径: src/client/nc/ui/{module}/{billCode}/{billCode}Controller.java
        String outputPath = String.format("%s/src/client/nc/ui/%s/%s/%sController.java",
                outputDir,
                billConfig.getModule(),
                billConfig.getClassNameLower(),
                billConfig.getClassName());

        writeToFile(outputPath, content);
        logger.info("生成Controller: {}", outputPath);
    }

    /**
     * 生成IPrivateBtn
     */
    private void generateIPrivateBtn(BillConfig billConfig) throws IOException {
        TemplateContext context = new TemplateContext();
        context.setBillConfig(billConfig);

        // 使用模板选择器获取模板路径
        String templatePath = TemplateSelector.getIPrivateBtnTemplate(billConfig);
        String content = templateEngine.render(templatePath, context);

        // 输出路径: src/client/nc/ui/{module}/{billCode}/{billCode}IPrivateBtn.java
        String outputPath = String.format("%s/src/client/nc/ui/%s/%s/%sIPrivateBtn.java",
                outputDir,
                billConfig.getModule(),
                billConfig.getClassNameLower(),
                billConfig.getClassName());

        writeToFile(outputPath, content);
        logger.info("生成IPrivateBtn: {}", outputPath);
    }


    /**
     * 生成ClientUI
     */
    private void generateClientUI(BillConfig billConfig) throws IOException {
        TemplateContext context = new TemplateContext();
        context.setBillConfig(billConfig);

        // 使用模板选择器获取模板路径
        String templatePath = TemplateSelector.getClientUITemplate(billConfig);
        String content = templateEngine.render(templatePath, context);

        // 输出路径: src/client/nc/ui/{module}/{billCode}/{billCode}ClientUI.java
        String outputPath = String.format("%s/src/client/nc/ui/%s/%s/%sClientUI.java",
                outputDir,
                billConfig.getModule(),
                billConfig.getClassNameLower(),
                billConfig.getClassName());

        writeToFile(outputPath, content);
        logger.info("生成ClientUI: {}", outputPath);
    }

    /**
     * 生成BusinessAction
     */
    private void generateBusinessAction(BillConfig billConfig) throws IOException {
        TemplateContext context = new TemplateContext();
        context.setBillConfig(billConfig);

        // 使用模板选择器获取模板路径
        String templatePath = TemplateSelector.getBusinessActionTemplate(billConfig);
        String content = templateEngine.render(templatePath, context);

        // 输出路径: src/client/nc/ui/{module}/{billCode}/{billCode}BusinessAction.java
        String outputPath = String.format("%s/src/client/nc/ui/%s/%s/%sBusinessAction.java",
                outputDir,
                billConfig.getModule(),
                billConfig.getClassNameLower(),
                billConfig.getClassName());

        writeToFile(outputPath, content);
        logger.info("生成BusinessAction: {}", outputPath);
    }

    /**
     * 生成Delegator
     */
    private void generateDelegator(BillConfig billConfig) throws IOException {
        TemplateContext context = new TemplateContext();
        context.setBillConfig(billConfig);

        // 使用模板选择器获取模板路径
        String templatePath = TemplateSelector.getDelegatorTemplate(billConfig);
        String content = templateEngine.render(templatePath, context);

        // 输出路径: src/client/nc/ui/{module}/{billCode}/{billCode}Delegator.java
        String outputPath = String.format("%s/src/client/nc/ui/%s/%s/%sDelegator.java",
                outputDir,
                billConfig.getModule(),
                billConfig.getClassNameLower(),
                billConfig.getClassName());

        writeToFile(outputPath, content);
        logger.info("生成Delegator: {}", outputPath);
    }

    /**
     * 生成EventHandler
     */
    private void generateEventHandler(BillConfig billConfig) throws IOException {
        TemplateContext context = new TemplateContext();
        context.setBillConfig(billConfig);

        // 使用模板选择器获取模板路径
        String templatePath = TemplateSelector.getEventHandlerTemplate(billConfig);
        String content = templateEngine.render(templatePath, context);

        // 输出路径: src/client/nc/ui/{module}/{billCode}/{billCode}EventHandler.java
        String outputPath = String.format("%s/src/client/nc/ui/%s/%s/%sEventHandler.java",
                outputDir,
                billConfig.getModule(),
                billConfig.getClassNameLower(),
                billConfig.getClassName());

        writeToFile(outputPath, content);
        logger.info("生成EventHandler: {}", outputPath);
    }

    /**
     * 生成RefModel
     */
    private void generateRefModel(BillConfig billConfig) throws IOException {
        TemplateContext context = new TemplateContext();
        context.setBillConfig(billConfig);

        // 使用模板选择器获取模板路径
        String templatePath = TemplateSelector.getRefModelTemplate(billConfig);
        String content = templateEngine.render(templatePath, context);

        // 输出路径: src/client/nc/ui/{module}/{billCode}/{billCode}RefModel.java
        String outputPath = String.format("%s/src/client/nc/ui/%s/%s/%sRefModel.java",
                outputDir,
                billConfig.getModule(),
                billConfig.getClassNameLower(),
                billConfig.getClassName());

        writeToFile(outputPath, content);
        logger.info("生成RefModel: {}", outputPath);
    }

    /**
     * 写入文件
     */
    private void writeToFile(String filePath, String content) throws IOException {
        File file = new File(filePath);
        File parentDir = file.getParentFile();

        if (!parentDir.exists()) {
            parentDir.mkdirs();
        }

        FileUtils.writeStringToFile(file, content, "GBK");
    }
}

File: src/core/generators/ClientGenerator.ts
import type { BillConfig } from '../models';
import TemplateEngine from '../utils/templateEngine';
import path from 'path';

export class ClientGenerator {
  private templateEngine: TemplateEngine;
  private outputDir: string;

  constructor(outputDir: string) {
    this.outputDir = outputDir;
    this.templateEngine = new TemplateEngine();
  }

  async generate(billConfig: BillConfig): Promise<void> {
    console.log('开始生成客户端代码');

    await this.generateController(billConfig);
    await this.generateIPrivateBtn(billConfig);
    await this.generateClientUI(billConfig);
    await this.generateBusinessAction(billConfig);
    await this.generateDelegator(billConfig);
    await this.generateEventHandler(billConfig);
    await this.generateRefModel(billConfig);

    console.log('客户端代码生成完成');
  }

  private async generateController(billConfig: BillConfig): Promise<void> {
    let templatePath: string;
    if (billConfig.billType === 'archive') {
      templatePath = 'client/by-type/archive/Controller_Archive.ejs';
    } else if (billConfig.billType === 'multi') {
      templatePath = 'client/by-type/multi/Controller_Multi.ejs';
    } else {
      templatePath = 'client/by-type/single/Controller_Single.ejs';
    }

    const context = { bill: billConfig, date: new Date().toISOString().split('T')[0] };
    const outputPath = path.join(
      this.outputDir,
      `src/client/nc/ui/${billConfig.packageName}/${billConfig.billCode.toLowerCase()}/controller/ClientCtrl.java`
    );
    
    await this.templateEngine.renderAndWrite(templatePath, outputPath, context);
    console.log(`生成Controller: ${outputPath}`);
  }

  private async generateIPrivateBtn(billConfig: BillConfig): Promise<void> {
    const context = { bill: billConfig, date: new Date().toISOString().split('T')[0] };
    const templatePath = 'client/common/IPrivateBtn.ejs';
    const outputPath = path.join(
      this.outputDir,
      `src/client/nc/ui/${billConfig.packageName}/${billConfig.billCode.toLowerCase()}/controller/IPrivateBtn.java`
    );
    
    await this.templateEngine.renderAndWrite(templatePath, outputPath, context);
    console.log(`生成IPrivateBtn: ${outputPath}`);
  }

  private async generateClientUI(billConfig: BillConfig): Promise<void> {
    let templatePath: string;
    if (billConfig.billType === 'archive') {
      templatePath = 'client/by-type/archive/ClientUI_Archive.ejs';
    } else if (billConfig.billType === 'multi') {
      templatePath = 'client/by-type/multi/ClientUI_Multi.ejs';
    } else {
      templatePath = 'client/by-type/single/ClientUI_Single.ejs';
    }

    const context = { bill: billConfig, date: new Date().toISOString().split('T')[0] };
    const outputPath = path.join(
      this.outputDir,
      `src/client/nc/ui/${billConfig.packageName}/${billConfig.billCode.toLowerCase()}/ui/ClientUI.java`
    );
    
    await this.templateEngine.renderAndWrite(templatePath, outputPath, context);
    console.log(`生成ClientUI: ${outputPath}`);
  }

  private async generateBusinessAction(billConfig: BillConfig): Promise<void> {
    const context = { bill: billConfig, date: new Date().toISOString().split('T')[0] };
    const templatePath = 'client/common/BusinessAction.ejs';
    const outputPath = path.join(
      this.outputDir,
      `src/client/nc/ui/${billConfig.packageName}/${billConfig.billCode.toLowerCase()}/action/BusinessAction.java`
    );
    
    await this.templateEngine.renderAndWrite(templatePath, outputPath, context);
    console.log(`生成BusinessAction: ${outputPath}`);
  }

  private async generateDelegator(billConfig: BillConfig): Promise<void> {
    const context = { bill: billConfig, date: new Date().toISOString().split('T')[0] };
    const templatePath = 'client/common/Delegator.ejs';
    const outputPath = path.join(
      this.outputDir,
      `src/client/nc/ui/${billConfig.packageName}/${billConfig.billCode.toLowerCase()}/delegator/ClientDelegator.java`
    );
    
    await this.templateEngine.renderAndWrite(templatePath, outputPath, context);
    console.log(`生成Delegator: ${outputPath}`);
  }

  private async generateEventHandler(billConfig: BillConfig): Promise<void> {
    const context = { bill: billConfig, date: new Date().toISOString().split('T')[0] };
    const templatePath = 'client/common/EventHandler.ejs';
    const outputPath = path.join(
      this.outputDir,
      `src/client/nc/ui/${billConfig.packageName}/${billConfig.billCode.toLowerCase()}/handler/EventHandler.java`
    );
    
    await this.templateEngine.renderAndWrite(templatePath, outputPath, context);
    console.log(`生成EventHandler: ${outputPath}`);
  }

  private async generateRefModel(billConfig: BillConfig): Promise<void> {
    const context = { bill: billConfig, date: new Date().toISOString().split('T')[0] };
    const templatePath = 'client/common/RefModel.ejs';
    const outputPath = path.join(
      this.outputDir,
      `src/client/nc/ui/${billConfig.packageName}/${billConfig.billCode.toLowerCase()}/refmodel/RefModel.java`
    );
    
    await this.templateEngine.renderAndWrite(templatePath, outputPath, context);
    console.log(`生成RefModel: ${outputPath}`);
  }
}

export default ClientGenerator;
